### DISCUSSION TITLE
Include Usage for Non-id based fields

### DISCUSSION THREAD

**Discussion by CyberCyclone** [ravendb/ravendb#13831] [Q&A] [answered]
I have two collections.

Blocks:
```
{
    "context": {
        "block_hash": "f988da2289b62f878f1417eeb3ceb7c5278addc958886ee8d8ad6b92cc449a9d",
        "block_number": 5837882,
        "slot": 31867205
    },
    "block": {
        "era": "Mary",
        "body_size": 15408,
        "issuer_vkey": "867bfa92c9ff312ef4a5553c770768f21e7729b5d2a8a16e983c8f27ec586d07",
        "tx_count": 19,
        "slot": 31867205,
        "hash": "f988da2289b62f878f1417eeb3ceb7c5278addc958886ee8d8ad6b92cc449a9d",
        "number": 5837882,
        "previous_hash": "4a87a4545256f542770ba1e5d3e05f05f555efa26aada81227fa2dfd317cabff"
    },
    "timestamp": 1623433496000,
    "@metadata": {
        "@collection": "Block"
    }
}
```

Transactions:

```
{
    "context": {
        "block_hash": "f988da2289b62f878f1417eeb3ceb7c5278addc958886ee8d8ad6b92cc449a9d",
        "block_number": 5837882,
        "slot": 31867205,
        "tx_idx": 0,
        "tx_hash": "34f38a0189f3f7f857c99e87bafcb9d7a04d8afa9ea3a43cd7e5b70e2cf8b303"
    },
    "transaction": {
        "hash": "34f38a0189f3f7f857c99e87bafcb9d7a04d8afa9ea3a43cd7e5b70e2cf8b303",
        "fee": 168449,
        "ttl": 31874339,
        "input_count": 1,
        "output_count": 2,
        "mint_count": 0,
        "total_output": 26657831551
    },
    "timestamp": 1623433496000,
    "@metadata": {
        "@collection": "Transaction"
    }
}
```

A transaction document is related to the block document by `transaction.context.block_hash` = `block.block.hash` and a block document will be a one to many relationship.

I'm trying to do a query with the `Include` statement however from the documentation it appears that the `Include` statement only acts on the Id field of the related document. Is this true or is there some undocumented syntax to make `Include`  match based on document fields?

I've also looked into creating an Index but the `LoadDocument` also appears to only act on Ids.

Basically what I'm after is a MongoDB style of the $lookup aggregation command because I don't see how I can use this structure with RavenDB without it since I need to do queries in both directions. E.g, When querying a Block doc, get all Transaction docs, or When querying a Transaction doc, get the Block doc.

Thanks!

**Comment by arekpalinski** [comment]
You're correct. Include and LoadDocument expect to get a property which stores a document ID. I assume the hash is unique. Can you make it part of ID for documents in Blocks collections e.g. `blocks/f988da2289b62f878f1417eeb3ceb7c5278addc958886ee8d8ad6b92cc449a9d`?

**Comment by ayende** [comment, marked as answer]
There is some flexibility in `include` in terms of adding postfix / prefix values, but in general, it always operates on the id of the related document.
This is important for performance reasons, since otherwise you can get into some pretty costly situations.

In your case, `block` I assume is using something like `block/$blockHash` for the id? If so, you can use the prefix syntax like this:

```
from Transactions
where transaction.total_output > $value
include '(blocks/)context.block_hash'
```

That would give you the relevant transactions and the associated blocks for them.

Going the other way, you may have an unlimited number of values, mind. Typically, if you want to have it both ways, you'll store the `transactions` in the block as well. 

Alternatively, depending on the size / usage, you'll put everything in a single document.

**Reply by CyberCyclone** [reply]
Thanks for that @ayende and @arekpalinski. It would definitely need to go both ways. The query that @ayende gave for Transaction works for loading the parent Block for a Transaction, however if I wanted to get all the Transactions for a Block document then I would need to add all the Transactions which be a lot of duplication. 

Since there's 13 other document types (other than Transactions) that a Block document can contain, it does make sense to simply store them under the single Block document. The current max Block document size that would be stored is limited to 82KB and will physically never get to the max document limit that RavenDB can hold for a single document. So I'll go with storing it in a single document.

**Reply by mcrio** [reply]
Hi @ayende is that possible using C# `IRavenQueriable`. 

I was hoping something like this would work but it doesn't:
`.Include(document => "users/" + document.TeanantId + "/" + document.UserId)`

Asking as I am building the query based on different search criteria.

A solution would be to have an additional property store that composed id, but it seems like wasted space : )

**Reply by georgiosd** [reply]
@CyberCyclone use a semantic key and load using `LoadStartsWith` instead, same way you can query storage on Substrate.
`blocks/<block hash>/<transaction hash>` or `blocks/<block hash>/transactions/<transaction hash>` - the latter gives you more flexibility later on if you want to add more "child" keys.

**Reply by ayende** [reply]
From the C# API, you may be able to do something like:

```
 from tx in session.Query<Transaction>()
 let include = (doc) => RavenQuery.Include( "users/" + doc.TenantId + "/" + doc.UserId)
 let _ = include(tx)
 where tx.TotalOutput > foo
 select tx
```

Notepad code, mind.

You can certain do that as a JS query, however.

**Reply by mcrio** [reply]
Thanks @ayende  There doesn't seem to be a `RavenQuery.Include()` available but I'll dig deeper.  
p.s. I'm building a query based on search criteria so trying to avoid raw queries because of proper parameter escaping and cleaner code.

**Reply by mcrio** [reply]
Btw any chance allowing simple expressions like these `.Include(document => "users/" + document.TeanantId + "/" + document.UserId)` would be possible to implement?

**Reply by ayende** [reply]
That is something that you can already do via a JS invocation. 

```
declare function output(i) {
        include('users/' + i.UserId + '/' + i.TenantId);
        return i;
}
from 'Items' as i select output(i)
```

It isn't easy to get that to work via the Linq proivder, however.
